### ISSUE TITLE
parallel even with parallel disabled

### ISSUE THREAD

**Issue by fommil** [sbt/sbt#1886]
## original report

I have set

```
    parallelExecution in Test := false,
    parallelExecution in IntegrationTest := false,
    testForkedParallel in Test := false,
    testForkedParallel in IntegrationTest := false,
```

(this is actually environment specific, i.e. disable parallel runs on travis, but allow locally)

but ScalaTest WordSpecs that have the `ParallelTestExecution` mixin are still running out of order and apparently in parallel. e.g. https://travis-ci.org/ensime/ensime-server/builds/52494180

**Comment by eed3si9n** [comment]
This is super confusing because there are two layers of parallelism and multiple semantics associated with the word "test".
1. In sbt, parallelism refers to running more than one task in parallel, and `Test` refers to the `Test` configuration.
2. With in a particular testing framework like ScalaTest and Specs2, parallelism probably refers to running more than one example/assertion/check in parallel, and the word test refers to the assertion checking.

Could you rephrase your case in terms of steps, problems, and expectations, so I can reproduce this?

**Comment by fommil** [comment]
I don't really know how to simplify this any further. This isn't just something that I discovered after making a couple of steps: I've just spotted some weird behaviour.

In my project https://github.com/ensime/ensime-server/ we have several modules, and in each module we have tests and integration tests. Some of those tests are marked as `ParallelTestExecution` and some of them are not.

When we're running in travis, we want to turn off all parallelisation because their machines aren't powerful enough. However, even when I turn off everything to do with parallel strategies in sbt, the tests appear to run out of order. It's like each test is being thrown into a hashmap and evaluated in a random order. It would be a lot cleaner if everything ran in order, as it means that the subtests would match with their heading.

I'm willing to tolerate out-of-order execution of tests when I get the benefits of parallelism, but this is just confusing.

**Comment by fommil** [comment]
is this maybe scalatest doing the shuffling?

**Comment by fommil** [comment]
I'll try to explain what I'm seeing in the build: let's say we have WordSpecs (top level in this hierachy) with tests underneat them:
- MyThing
  - should foo
  - should bar
  - should baz
- MyOtherThing
  - should paper
  - should scissor
  - should stone

when I run the tests, they might be executed and output like this (even though all parallelisation is turned off)
- MyThing
- MyOtherThing
  - should baz
  - should scissor
  - should foo
  - should paper
  - should stone
  - should bar

That's what we're seeing in the travis link above.

**Comment by eed3si9n** [comment]
ok, so we have three parallel semantics.
1. task parallelism within a single command execution
2. test parallelism within a single project (`parallelExecution in Test := false`)
3. example parallelism within a single test suite (`ParallelTestExecution`)

So to minimize your situation,

## steps
1. make a multi-project build with subproject `sexpress` and `server`
2. each subproject runs in a loop displaying "sexpress success" (or "server success") every second for a minute
3. run `sbt test`

## problem

The output is interleaved.

## expectation

The output is not interleaved.

## note

See http://www.scala-sbt.org/0.13/docs/Parallel-Execution.html

> To restrict the number of concurrently executing tests in all projects, use:

``` scala
concurrentRestrictions in Global += Tags.limit(Tags.Test, 1)
```

**Comment by fommil** [comment]
Aah, cool thanks! It seems bizarre that I have to explicitly disable an experimental feature.

**Comment by eed3si9n** [comment]
The experimental part I think is the fact that you can tweak the task parallelism within a single command execution.
